### PR TITLE
Allow embedders to specify software renderer configs.

### PR DESCRIPTION
**Pull request by chinmaygarde**
This updates the embedder API but does not make any breaking API/ABI changes.

### CODE REVIEW

File: shell/platform/embedder/embedder.cc [hunk 1]
```diff
@@ -56,76 +56,68 @@ bool IsRendererValid(const FlutterRendererConfig* config) {
...
+    return false;
   }
 
-  if (args == nullptr) {
-    return kInvalidArguments;
+  return true;
+}
+
+static bool IsRendererValid(const FlutterRendererConfig* config) {
+  if (config == nullptr || config->type != kOpenGL) {
```
Comment (cbracken) [inline shell/platform/embedder/embedder.cc]: Given the checks below, why the `config->type != kOpenGL`?

File: shell/platform/embedder/embedder_surface_software.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#include "flutter/shell/gpu/gpu_surface_software.h"
+#include "flutter/shell/platform/embedder/embedder_surface.h"
+
+namespace shell {
+
+class EmbedderSurfaceSoftware final : public EmbedderSurface,
+                                      public GPUSurfaceSoftwareDelegate {
+ public:
+  struct SoftwareDispatchTable {
+    std::function<bool(const void* allocation, size_t row_bytes, size_t height)>
```
Comment (jason-simmons) [inline shell/platform/embedder/embedder_surface_software.h]: Add width and pixel format arguments to this callback

Comment (chinmaygarde) [reply shell/platform/embedder/embedder_surface_software.h]: I don't want to make the pixel format configurable as that includes a fair bit of complexity at the API level. So all allocations will be N32. Towards that end, with width can be inferred from the row bytes.

File: shell/platform/embedder/embedder.cc (hunk 1 above)
Comment (chinmaygarde) [reply shell/platform/embedder/embedder.cc]: I wanted the method to be self contained.

File: shell/platform/embedder/embedder_surface_software.h (hunk 2 above)
Comment (jason-simmons) [reply shell/platform/embedder/embedder_surface_software.h]: I don't think the format needs to be configurable.  But if this is a callback supplied by the user, then I'd like to reserve the flexibility to change the pixel format or add padding in the bitmap given to the callback.

If the bitmap will always be N32 with no padding, then document that in embedder.h

Comment (chinmaygarde) [reply shell/platform/embedder/embedder_surface_software.h]: I had documented the pixel format and also added a not about buffer ownership.

About padding: Though it is unlikely that we will ever need to change the padding the buffer, changing the padding but keeping the pixel format the same will result in a change in the row bytes. So I believe the API is its current state is adequate.

**Comment by jason-simmons** [comment]
LGTM

**Pull request merged**